satellite/repair/repairer/ec.go

```diff
@@ -0,0 +1,468 @@
+				successfulPieces++
+
+				return
+			}
+		})
+	}
+
+	limiter.Wait()
+
+	if successfulPieces < es.RequiredCount() {
+		return nil, Error.New("couldn't download enough pieces, number of successful downloaded pieces (%d) is less than required number (%d)", successfulPieces, es.RequiredCount())
+	}
+
+	fec, err := infectious.NewFEC(es.RequiredCount(), es.TotalCount())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	esScheme := eestream.NewUnsafeRSScheme(fec, es.ErasureShareSize())
+	expectedSize := pieceSize * int64(es.RequiredCount())
+
+	decodeReader := eestream.DecodeReaders(ctx, cancel, ec.log.Named("decode readers"), pieceReaders, esScheme, expectedSize, 0, false)
+
+	return decodeReader, nil
+}
+
+// downloadAndVerifyPiece downloads a piece from a storagenode,
+// expects the original order limit to have the correct piece public key,
+// and expects the hash of the data to match the signed hash provided by the storagenode.
+func (ec *ECRepairer) downloadAndVerifyPiece(ctx context.Context, limit *pb.AddressedOrderLimit, privateKey storj.PiecePrivateKey, pieceSize int64, bufferSize int) (data []byte, err error) {
+	// contact node
+	ps, err := ec.dialPiecestore(ctx, &pb.Node{
+		Id:      limit.GetLimit().StorageNodeId,
+		Address: limit.GetStorageNodeAddress(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	downloader, err := ps.Download(ctx, limit.GetLimit(), privateKey, 0, pieceSize)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, downloader.Close()) }()
+
+	var calculatedHash []byte
+	pieceBytes := make([]byte, 0, pieceSize)
+	buffer := make([]byte, bufferSize)
+	newHash := pkcrypto.NewHash()
+
+	for {
+		// download full piece
+		n, readErr := downloader.Read(buffer)
+		if readErr == io.EOF {
+			calculatedHash = newHash.Sum(buffer[:n])
+			break
+		}
+		if readErr != nil {
+			return nil, readErr
+		}
+
+		// add new data to hash calculation
+		_, _ = newHash.Write(buffer[:n]) // guaranteed not to return an error
+		// add new data to piece bytes
+		pieceBytes = append(pieceBytes, buffer[:n]...)
+	}
+
+	if int64(len(pieceBytes)) != pieceSize {
+		return nil, Error.New("didn't download the correct amount of data, want %d, got %d", pieceSize, len(pieceBytes))
+	}
+
+	// get signed piece hash and original order limit
+	hash, originalLimit := downloader.GetHashAndLimit()
+	if hash == nil {
+		return nil, Error.New("Hash was not sent from storagenode.")
+	}
+	if originalLimit == nil {
+		return nil, Error.New("Original order limit was not sent from storagenode.")
+	}
+
+	// verify the hashes from storage node
+	if err := verifyPieceHash(ctx, originalLimit, hash, calculatedHash); err != nil {
+		return nil, err
+	}
+
+	// verify order limit from storage node is signed by the satellite
+	if err := verifyOrderLimitSignature(ctx, ec.satelliteSignee, originalLimit); err != nil {
```

Comment:
`verifyOrderLimitSignature` should be before `verifyPieceHash` for cases when `limit.UplinkPublicKey` was manipulated